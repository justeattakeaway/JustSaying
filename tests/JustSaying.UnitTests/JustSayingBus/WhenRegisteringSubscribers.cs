using Amazon.SQS.Model;
using JustSaying.TestingFramework;
using JustSaying.UnitTests.Messaging.Channels.SubscriptionGroupTests;
using Newtonsoft.Json;

namespace JustSaying.UnitTests.JustSayingBus;

public sealed class WhenRegisteringSubscribers(ITestOutputHelper outputHelper) : GivenAServiceBus(outputHelper), IDisposable
{
    private FakeSqsQueue _queue1;
    private FakeSqsQueue _queue2;
    private CancellationTokenSource _cts;

    protected override void Given()
    {
        base.Given();

        static IEnumerable<Message> GetMessages(CancellationToken cancellationToken)
        {
            while (!cancellationToken.IsCancellationRequested)
            {
                yield return new TestMessage();
            }
        }

        _queue1 = new FakeSqsQueue(ct =>Task.FromResult(GetMessages(ct)), "queue1");
        _queue2 = new FakeSqsQueue(ct => Task.FromResult(GetMessages(ct)), "queue2");
    }

    protected override async Task WhenAsync()
    {
        SystemUnderTest.AddMessageMiddleware<OrderAccepted>(_queue1.QueueName,
            new InspectableMiddleware<OrderAccepted>());
        SystemUnderTest.AddMessageMiddleware<OrderRejected>(_queue1.QueueName,
            new InspectableMiddleware<OrderRejected>());
        SystemUnderTest.AddMessageMiddleware<SimpleMessage>(_queue1.QueueName,
            new InspectableMiddleware<SimpleMessage>());

        SystemUnderTest.AddQueue("groupA", _queue1);
        SystemUnderTest.AddQueue("groupB", _queue2);

        _cts = new CancellationTokenSource();
        _cts.CancelAfter(TimeSpan.FromSeconds(5));

        await SystemUnderTest.StartAsync(_cts.Token);
    }

    [Fact]
    public async Task SubscribersStartedUp()
    {
        await Patiently.AssertThatAsync(OutputHelper,
            () =>
            {
                _queue1.ReceiveMessageRequests.Count.ShouldBeGreaterThan(0);
                _queue2.ReceiveMessageRequests.Count.ShouldBeGreaterThan(0);
            });
    }

    [Fact]
    public void AndInterrogationShowsSubscribersHaveBeenSet()
    {
        dynamic response = SystemUnderTest.Interrogate();

        string json = JsonConvert.SerializeObject(response.Data.Middleware.Data.Middlewares, Formatting.Indented);

        json.ShouldMatchApproved(c => c.SubFolder("Approvals"));
    }

    private class TestMessage : Message
    {
        public TestMessage()
        {
            Body = "TestMessage";
        }
    }

    public void Dispose()
    {
        _cts?.Dispose();
    }
}
