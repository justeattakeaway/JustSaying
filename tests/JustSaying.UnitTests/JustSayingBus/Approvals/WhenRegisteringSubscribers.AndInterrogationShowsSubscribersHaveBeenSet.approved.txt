[
  {
    "MessageType": "OrderAccepted",
    "QueueName": "queue1",
    "MiddlewareChain": [
      "InspectableMiddleware`1[OrderAccepted]"
    ]
  },
  {
    "MessageType": "OrderRejected",
    "QueueName": "queue1",
    "MiddlewareChain": [
      "InspectableMiddleware`1[OrderRejected]"
    ]
  },
  {
    "MessageType": "SimpleMessage",
    "QueueName": "queue1",
    "MiddlewareChain": [
      "InspectableMiddleware`1[SimpleMessage]"
    ]
  }
]